fetch/fetch.go

```diff
@@ -650,3 +706,45 @@ func (f *Fetch) GetHash(hash types.Hash32, h Hint, validateHash bool) chan HashD
 
 	return resChan
 }
+
+func (f *Fetch) mapPeerToHash(hash types.Hash32, peer p2p.Peer) {
+	f.hashToPeersM.Lock()
+	defer f.hashToPeersM.Unlock()
+
+	peers, exists := f.hashToPeers.Get(hash)
```

Comment:
we should hide this logic in cache. i.e. hashToPeers.Add() should check its internal cache first before adding.
and you should change the hashToPeers.Add() to take a set of peers. otherwise we will get the same hash from cache `len(peers)` times.

Reply:
logic hidden, cheers